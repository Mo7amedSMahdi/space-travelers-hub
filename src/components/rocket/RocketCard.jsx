/* eslint-disable object-curly-newline */
/* eslint-disable react/jsx-one-expression-per-line */
import { PropTypes } from "prop-types";
import { Container, Row, Col, Image, Button, Badge } from "react-bootstrap";

const RocketCard = ({
  rocketName,
  description,
  flckrImg,
  reserved,
  reserve,
  cancel,
}) => (
  <li className="py-3 my-3">
    <Container>
      <Row>
        <Col xs lg={3}>
          <Image
            src={flckrImg}
            width="100%"
            height="auto"
            rounded
            alt="image"
          />
        </Col>
        <Col>
          <h2>{rocketName}</h2>
          <p>
            <span>
              {reserved ? (
                <Badge pill bg="info">
                  Reserved
                </Badge>
              ) : (
                ""
              )}
            </span>{" "}
            {description}
          </p>
          <Button
            variant={reserved ? "light" : "primary"}
            onClick={reserved ? cancel : reserve}
          >
            {reserved ? "Cancel Rocket" : "Reserve Rocket"}
          </Button>
        </Col>
      </Row>
    </Container>
  </li>
);

RocketCard.propTypes = {
  rocketName: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  flckrImg: PropTypes.string.isRequired,
  reserved: PropTypes.bool,
  reserve: PropTypes.func,
  cancel: PropTypes.func,
};

RocketCard.defaultProps = {
  reserve: null,
  cancel: null,
  reserved: false,
};

export default RocketCard;
