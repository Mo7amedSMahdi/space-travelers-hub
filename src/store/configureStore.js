import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';
import missionsReducer from './missions';

const rootReducer = combineReducers({ missions: missionsReducer });

const store = configureStore({ reducer: rootReducer });

export default store;
