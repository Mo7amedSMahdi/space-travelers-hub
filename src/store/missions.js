import { createSlice } from '@reduxjs/toolkit';
import moment from 'moment';
// import axios from 'axios';
import { apiCallBegan } from './api';

const url = '/missions';
const slice = createSlice({
  name: 'missions',
  initialState: {
    list: [],
    loading: false,
    lastFetch: null,
  },
  reducers: {
    missionsRequested: (mission) => {
      mission.loading = true;
    },
    missionsRecieved: (mission, action) => {
      mission.list = action.payload.map((m) => ({
        id: m.mission_id,
        name: m.mission_name,
        description: m.description,
        reserved: false,
      }));
      mission.loading = false;
      mission.lastFetch = Date.now();
    },
    missionsRequestFailed: (mission) => {
      mission.loading = false;
    },
    joinMission: (mission, action) => {
      const index = mission.list.findIndex((m) => m.id === action.payload);
      mission.list[index].reserved = true;
    },
  },
});

export const {
  missionsRequested, missionsRecieved, missionsRequestFailed, joinMission,
} = slice.actions;
export default slice.reducer;

// const missionsRecieved = createAction('missionsRecieved');

export const getMissions = () => async (dispatch, getState) => {
  const { lastFetch } = getState().missions;

  const diffInMinutes = moment().diff(moment(lastFetch), 'minutes');
  if (diffInMinutes < 10) return;
  dispatch(
    apiCallBegan({
      url,
      onStart: missionsRequested.type,
      onSuccess: missionsRecieved.type,
      onError: missionsRequestFailed.type,
    }),
  );
};
