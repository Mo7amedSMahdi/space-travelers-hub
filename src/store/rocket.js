import { createSlice } from "@reduxjs/toolkit";
import moment from "moment";
import { apiCallBegan } from "./api";

const url = "/rockets";
const slice = createSlice({
  name: "rockets",
  initialState: {
    list: [],
    loading: false,
    lastFetch: null,
  },
  reducers: {
    rocketsRequested: (rockets) => {
      rockets.loading = true;
    },
    rocketsRecieved: (rockets, action) => {
      rockets.list = action.payload;
      rockets.loading = false;
      rockets.lastFetch = Date.now();
    },
    rocketsRequestFailed: (rocket) => {
      rocket.loading = false;
    },
    joinRocket: (rocket, action) => {
      const index = rocket.list.findIndex((m) => m.id === action.payload);
      rocket.list[index].reserved = true;
    },
    leaveRocket: (rocket, action) => {
      const index = rocket.list.findIndex((m) => m.id === action.payload);
      rocket.list[index].reserved = false;
    },
  },
});

export const {
  rocketsRequested,
  rocketsRecieved,
  rocketsRequestFailed,
  joinRocket,
  leaveRocket,
} = slice.actions;
export default slice.reducer;
// const missionsRecieved = createAction('missionsRecieved');
export const getRockets = () => async (dispatch, getState) => {
  const { lastFetch } = getState().missions;
  const diffInMinutes = moment().diff(moment(lastFetch), "minutes");
  if (diffInMinutes < 10) return;
  dispatch(
    apiCallBegan({
      url,
      onStart: rocketsRequested.type,
      onSuccess: rocketsRecieved.type,
      onError: rocketsRequestFailed.type,
    }),
  );
};
