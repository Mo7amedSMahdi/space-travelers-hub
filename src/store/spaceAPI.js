class spaceAPI {
  static baseURL = 'https://api.spacexdata.com/v3';

  //   fetch missions
  static getMissions = async () => fetch(`${this.baseURL}/missions`).then(async (res) => {
    if (res.ok) {
      try {
        const data = await res.json();
        return data;
      } catch (e) {
        return [];
      }
    }

    return null;
  });

  // fetch rockets
  static getRockets = async () => fetch(`${this.baseURL}/rockets`).then(async (res) => {
    if (res.ok) {
      try {
        const data = await res.json();
        return data;
      } catch (e) {
        return [];
      }
    }

    return null;
  });
}

export default spaceAPI;
