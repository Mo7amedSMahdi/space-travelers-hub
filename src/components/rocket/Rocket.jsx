import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import { getRockets, joinRocket, leaveRocket } from '../../store/rocket';
import RocketCard from './RocketCard';

const Rockets = () => {
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets);

  useEffect(() => {
    dispatch(getRockets());
  }, []);

  return (
    <>
      {/* TODO: Header here */}

      <Container className="px-5">
        <Row>
          <Col>
            <ul className="list-unstyled">
              {rockets.list.map((rocket) => (
                <RocketCard
                  key={rocket.id}
                  rocketName={rocket.rocket_name}
                  description={rocket.description}
                  flckrImg={rocket.flickr_images[0]}
                  reserved={rocket.reserved}
                  reserve={() => dispatch(joinRocket(rocket.id))}
                  cancel={() => dispatch(leaveRocket(rocket.id))}
                />
              ))}
            </ul>
          </Col>
        </Row>
      </Container>
    </>
  );
};

export default Rockets;
