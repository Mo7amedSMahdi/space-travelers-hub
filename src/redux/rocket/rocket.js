import spaceAPI from "../../api/spaceAPI";

// actions
const FETCH_ROCKET = "FETCH_ROCKET";
const RESERVE_ROCKET = "RESERVE_MISSION";
const CANCEL_ROCKET = "CANCEL_ROCKET";

// fetch rockets
export const fetchRockets = () => async (dispatch) => {
  const rockets = await spaceAPI.getRockets();

  if (rockets) {
    dispatch({ type: FETCH_ROCKET, rockets });
  }
};

// TODO: reducer to go here
export default function reducer(state = [], action) {
  switch (action.type) {
    case FETCH_ROCKET:
      return action.rockets;

    case RESERVE_ROCKET:
      return state.map((rocket) => {
        if (rocket.id !== action.id) return rocket;
        return { ...rocket, reserved: true };
      });

    case CANCEL_ROCKET:
      return state.map((rocket) => {
        if (rocket.id !== action.id) return rocket;
        return { ...rocket, reserved: false };
      });

    default:
      return state;
  }
}

// action creators
export function reserveRocket({ id }) {
  return async (dispatch) => {
    dispatch({
      type: RESERVE_ROCKET,
      id,
    });
  };
}

export function cancelRocket({ id }) {
  return async (dispatch) => {
    dispatch({
      type: CANCEL_ROCKET,
      id,
    });
  };
}
