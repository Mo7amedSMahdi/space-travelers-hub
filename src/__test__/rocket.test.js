import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import Rockets from "../components/rocket/Rocket";
import store from "../store/configureStore";
import { rocketsRequested, joinRocket, leaveRocket } from "../store/rocket";

describe("Rockets tests", () => {
  it("render test", () => {
    const tree = render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it("should return an action with type rockets/rocketsRequested", () => {
    const action = rocketsRequested.type;
    expect(action).toBe("rockets/rocketsRequested");
  });

  it("should return an action with type rockets/joinRocket", () => {
    const action = joinRocket.type;
    expect(action).toBe("rockets/joinRocket");
  });

  it("should should return an action with type rockets/leaveRocket", () => {
    const action = leaveRocket.type;
    expect(action).toBe("rockets/leaveRocket");
  });
});
