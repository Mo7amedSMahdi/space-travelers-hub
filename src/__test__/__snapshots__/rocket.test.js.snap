// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets tests render test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="px-5 container"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <ul
              class="list-unstyled"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="px-5 container"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <ul
            class="list-unstyled"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
